### PR TITLE
n-dimensional batch normalization (addresses #29)

### PR DESCRIPTION
**Pull request by IanCoolidge0**
This PR adds a batch normalization layer in the format of other layers in MyNN. The layer is able to normalize batches of an arbitrary shape, given as the parameter `input_shape` upon creating the layer. The layer also stores moving averages of the computed means and variances, and can use these at test time by specifying `test=True` when calling the layer.

The layer has been tested in a dense network (based on the spiral dataset example notebook provided in MyNN). I created my own notebook which compares training speed of this network with and without batch norm (this is not on github right now). The only test I have not done yet is using a network with a higher dimensionality for `input_shape`; I plan to test it on an MNIST CNN soon. Right now statistics are just computed over axis 0 regardless, which I think is correct but can be tested. 

David edit: This addresses #29

### CODE REVIEW

**Comment by rsokl** [comment]
@IanCoolidge0 This is looking really good! Awesome documentation and code structure.

Regarding the computation of mean and variance, I think we can eek out improved performance by leveraging a neat trick. For the moment, let's treat `x` as a 1D array that we want to compute the mean and variance of. The way you compute the mean is great as-is, thus we have `E[x]`.  Now, [representing variance as](https://en.wikipedia.org/wiki/Variance#Definition)
```
Var[x] = E[x^2] - E[x]^2
```
see that E[x^2] is computed most straightforwardly as `mean(x**2)`, which iterates over the elements of `x` twice and allocates an intermediate array of memory to store `x**2`. We can perform the same computation iterating over `x` once and avoiding the intermediate allocation via `np.dot(x, x) / len(x)`. In the 1D case, this can earn us a ~2x speedup over `mean(x**2)`.  

Of course, things are a bit more nuanced for us. We are dealing with tensors of (2+) dimensions, and we are only averaging over the batch dimension (axis-0). That is, we want to compute the following higher-dimensionality dot-product. 

<p align="center">
  
</p>

See that this notation indicates that `x` has at least 2 dimensions, as implied by the optional indices indicated by `(...)`. Some say that you would have to be a regular **Einstein** to compute this **summation** (hint hint :wink:)

Consider refactoring the way you compute variance in light of this discussion. Let me know what speedup you see, if any, by refactoring the implementation. You can use a shape-(100,3,32,32) random tensor for a sensible test case.

I'm happy to talk through how to compute that higher-dimensional dot-product via `einsum` if you find yourself stuck.

**Comment by IanCoolidge0** [comment]
I guess all the time I spent deciphering einsum during week 2 finally pays off :) It didn't take long once I realized it was just a dot product over `i` while essentially ignoring any trailing axes. Testing the actual result of the calculation got the same results as the old version.

I found a noticeable speedup, but it wasn't as much as I thought. I needed to raise the batch size to 10000 before it was really noticeable, and then the optimized calculation completed in ~600ms and the non-optimized version in ~900ms. I'm surprised that there wasn't at least a 2x speedup (based on what you mentioned above), but I guess there's probably some under-the-hood unpacking that goes on to make the original version faster.

Looking at the Pytorch layer, I think this now has essentially the same features. The only differences is that they put their epsilon parameter in the `__init__` instead, and they also have an `affine` parameter so the user can choose to not learn the gamma and beta parameters and just always normalize to zero mean and unit variance. If we need either of these things or anything else let me know, they should be simple enough to change.

**Comment by davidmascharka** [comment]
Sorry this has just been hanging here. I'll get around to reviewing and testing between tonight and Sunday.

File: mynn/layers/batchnorm.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+from mynn.initializers.constant import constant
+
+class batchnorm:
+    ''' A batch normalization layer.
+
+    This class will perform an n-dimensional batch normalization operation on an (N, D, ...)
+    shaped tensor scaled by γ of shape (D, ...) and shifted by β of shape (D, ...).
+    '''
+    def __init__(self, input_shape, momentum=0.1, gamma_initializer=constant,
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: For now, I don't think we need to give the initializers as a parameter. See the main body of my review.

File: mynn/layers/batchnorm.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    shaped tensor scaled by γ of shape (D, ...) and shifted by β of shape (D, ...).
+    '''
+    def __init__(self, input_shape, momentum=0.1, gamma_initializer=constant, 
+                 beta_initializer=constant, moving_mean_initializer=constant, 
+                 moving_var_initializer=constant, gamma_kwargs={"value": 1.0}, 
+                 beta_kwargs=None):
+        ''' Initialize a batch normalization layer.
+
+        Parameters
+        ----------
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: We'll only need `input_channels` and `momentum` here

File: mynn/layers/batchnorm.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            The function to use to initialize the moving variance tensor. 
+
+        gamma_kwargs : Optional[dictionary] (default={"value": 1.0})
+            The keyword arguments to pass to the gamma initialization function.
+            Initially set to pass a constant value of 1.0 to the scaling factor.
+
+        beta_kwargs : Optional[dictionary]
+            The keyword arguments to pass to the beta initialization function.
+        '''
+        self.gamma = gamma_initializer(*input_shape,
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: `self.gamma = mg.Tensor(np.ones((1, input_channels), dtype=np.float32)` should suffice for now

File: mynn/layers/batchnorm.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        gamma_kwargs : Optional[dictionary] (default={"value": 1.0})
+            The keyword arguments to pass to the gamma initialization function.
+            Initially set to pass a constant value of 1.0 to the scaling factor.
+
+        beta_kwargs : Optional[dictionary]
+            The keyword arguments to pass to the beta initialization function.
+        '''
+        self.gamma = gamma_initializer(*input_shape, 
+            **(gamma_kwargs if gamma_kwargs is not None else {}))
+        self.beta = beta_initializer(*input_shape,
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: `self.beta = mg.Tensor(np.zeros((1, input_channels), dtype=np.float32))` for the moment

File: mynn/layers/batchnorm.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            The keyword arguments to pass to the beta initialization function.
+        '''
+        self.gamma = gamma_initializer(*input_shape, 
+            **(gamma_kwargs if gamma_kwargs is not None else {}))
+        self.beta = beta_initializer(*input_shape, 
+            **(beta_kwargs if beta_kwargs is not None else {}))
+
+        self.beta.data = self.beta.data.astype(self.gamma.dtype)
+
+        self.moving_mean = moving_mean_initializer(*input_shape).data
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: The mean and variance here should just be `np.zeros` of the appropriate shape (which should be `(1, input_channels)`

File: mynn/layers/batchnorm.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    '''
+    def __init__(self, input_shape, momentum=0.1, gamma_initializer=constant, 
+                 beta_initializer=constant, moving_mean_initializer=constant, 
+                 moving_var_initializer=constant, gamma_kwargs={"value": 1.0}, 
+                 beta_kwargs=None):
+        ''' Initialize a batch normalization layer.
+
+        Parameters
+        ----------
+        input_shape : Sequence[int]
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: We don't need the full shape; we can do this with just having the number of input channels, which is much cleaner

File: mynn/layers/batchnorm.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            **(beta_kwargs if beta_kwargs is not None else {}))
+
+        self.beta.data = self.beta.data.astype(self.gamma.dtype)
+
+        self.moving_mean = moving_mean_initializer(*input_shape).data
+        self.moving_variance = moving_var_initializer(*input_shape).data
+    
+        self.momentum = momentum
+
+        self.input_shape = input_shape
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: Again, this can just be `input_channels` instead of a full shape

File: mynn/layers/batchnorm.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+    def __call__(self, x, epsilon=1e-5, test=False):
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: I don't like passing in an epsilon. Users should probably not mess with epsilon. Also, `1e-05` isn't very small and I don't know why the PyTorch people use that. I'd normally stick in `1e-08` myself. But whatever value it ends up being, it shouldn't be a parameter here, just add the number directly in the code below

File: mynn/layers/batchnorm.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            The data on which to perform batch normalization.
+
+        epsilon : float, optional (default=1e-5)
+            The value added to variance denominator to avoid division-by-zero errors.
+
+        test : boolean, optional (default=False)
+            Determines whether the layer is being used at training time or test time.
+            This determines which mean and variance are used for normalization.
+        '''
+        assert x.shape[1:] == self.input_shape, \
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: This can go away

File: mynn/layers/batchnorm.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            "Shape " + str(x.shape[1:]) + " of input tensor does not match required shape " + str(self.input_shape) + " for batch normalization."
+
+        if test:
+            x -= self.moving_mean
+            x /= np.sqrt(self.moving_variance + epsilon)
+
+            return self.gamma * x + self.beta
+        else:
+            batch_mean = mg.mean(x, axis=0)
+            batch_squared_mean = mg.einsum("i...,i...->...", x, x) / x.shape[0]
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: `batch_squared_mean = mg.einsum("ij...,ij...->j", x, x) / (x.size / x.shape[1])`

File: mynn/layers/batchnorm.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        assert x.shape[1:] == self.input_shape, \
+            "Shape " + str(x.shape[1:]) + " of input tensor does not match required shape " + str(self.input_shape) + " for batch normalization."
+
+        if test:
+            x -= self.moving_mean
+            x /= np.sqrt(self.moving_variance + epsilon)
+
+            return self.gamma * x + self.beta
+        else:
+            batch_mean = mg.mean(x, axis=0)
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: This should be over all the axes besides axis 1

File: mynn/layers/batchnorm.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+            batch_squared_mean = mg.einsum("i...,i...->...", x, x) / x.shape[0]
+
+            batch_variance = batch_squared_mean - batch_mean ** 2
+
+            self.moving_mean = (1 - self.momentum) * self.moving_mean + self.momentum * batch_mean.data
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: You can line-break all of these

```python
self.moving_mean *= (1 - self.momentum)
self.moving_mean += self.momentum * batch_mean.data
self.moving_variance *= (1 - self.momentum)
self.moving_variance += self.momentum * batch_variance.data
```

File: mynn/layers/batchnorm.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            self.moving_variance = (1 - self.momentum) * self.moving_variance + self.momentum * batch_variance.data
+
+            x -= batch_mean
+            x /= mg.sqrt(batch_variance + epsilon)
+
+            return self.gamma * x + self.beta
+
+    def reset_moving_statistics(self, moving_mean_initializer=constant,
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: I don't think we need this function

**Comment by rsokl** [comment]
@IanCoolidge0 I have been looking into the optimization issue (not seeing the factor of 2 speedup), and the issue is that, although the forward-pass is optimized, the autograd-generated backprop suffers because `mg.einsum('ij...,ij...->j', x, x)`, in its current implementation must backprop through `x` *twice* (once for each time it appears in the input). This obviously is highly undesirable. As such, I have implemented a patch for mygrad's einsum to identify such redundancies in its input and eliminate the need to compute the same gradient multiple times! This provides a substantial speedup. 

Looking further into this (I had time to kill on a train ride), I wrote out the analytic form for the derivatives of batch norm w.r.t (with respect to) x, gamma, and beta, and there are substantial optimizations that can be achieved by factoring the analytic form of the derivative.  I [have a branch in mygrad that implements the ND batchnorm op](https://github.com/rsokl/MyGrad/blob/batchnorm/mygrad/nnet/layers/batchnorm.py) - this gets a very nice speedup:

image

This is branched off MyGrad's 0.6 branch (named `topo_sort`), which still needs merging. Once it and this batchnorm branch is merged, you should use mygrad's batchnorm op in your layer here. I'll almost definitely have all of this material merged today (Monday).

Echoing David, your work is fantastic! In addition to providing the sorely-needed batchnorm layer, it also led me to realize that there were major improvements to be had in einsum's backprop!

**Comment by IanCoolidge0** [comment]
First off, I'm sorry you guys had to wade through my over-complicated (and somewhat incorrect) code. When I was reading the original batch norm paper I had a hunch that just directly computing batch statistics didn't work on a conv layer, but the PyTorch code through me off since the BatchNorm_3d, 4d, etc. classes all just appeared to be trivial subclasses of the same thing! I definitely missed something under the hood though, sorry about that.

I'll change this to use the mygrad batchnorm call once it's merged and remove all the unnecessary initializers. I'm glad some optimizations were able to come from this though!

**Comment by rsokl** [comment]
No worries at all! This is always part of the process. We all benefit a lot from the experience, as does the code. 
Also, welcome to the wonderful world of reading PyTorch's source code... :/

**Comment by rsokl** [comment]
@IanCoolidge0 the batchnorm op [is now available in mygrad](https://github.com/rsokl/MyGrad/pull/85)

**Comment by IanCoolidge0** [comment]
So while refactoring the layer to use the mygrad call, I've found a small issue with handling moving statistics. At test time, I just manually subtract and divide using the moving statistics (appropriately reshaped) since the calculation is trivial and doesn't need to be backpropped through anyway. At training time though, it's slightly more difficult because I need to somehow pull `mean` and `var` out of the batch norm call.

I see that the `BatchNorm` class stores these as instance variables when called, but when calling the op I never explicitly use the class (just the associated function) which doesn't give me a way to access those variables. I'm sure there's a correct way to get them without calculating them again, so I just want to make sure I'm not abusing the library when I do.

File: mynn/layers/batchnorm.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
+import numpy as np
+import mygrad as mg
+
+from mygrad.nnet.layers.batchnorm import batchnorm as tensor_batchnorm
+
+class batchnorm:
+    ''' A batch normalization layer.
+
+    This class will perform an n-dimensional batch normalization operation on an (N, D, ...)
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: Can you hyphenate this? "on an (N, D, ...)-shaped tensor"

File: mynn/layers/batchnorm.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+        self.moving_mean = np.zeros((1, input_channels), dtype=np.float32)
+        self.moving_variance = np.zeros((1, input_channels), dtype=np.float32)
+    
+        self.momentum = momentum
+
+        self.input_channels = input_channels
+
+    def __call__(self, x, test=False):
+        ''' Perform the forward-pass of n-dimensional batch normalization over axis 0 on `x`.
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: over axis 1

File: mynn/layers/batchnorm.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+        Parameters
+        ----------
+        x : Union[numpy.ndarray, mygrad.Tensor], shape=(N, D, ...)
+            The data on which to perform batch normalization.
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: I think "The data to normalize" is better wording here

File: mynn/layers/batchnorm.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+        test : boolean, optional (default=False)
+            Determines whether the layer is being used at training time or test time.
+            This determines which mean and variance are used for normalization.
```
Comment (davidmascharka) [inline mynn/layers/batchnorm.py]: I would add a little more exposition here than just saying "which mean and variance are used" for people who don't know what these different means are. Maybe just saying something like:

Whether to use averaged batch statistics or compute a mean and std over this batch

**Comment by rsokl** [comment]
@IanCoolidge0 you can access the mean and variance via the `BatchNorm` class. After you call `batchnorm`, the output tensor has the attribute `creator` - this is a reference to the `BatchNorm` instance that was used to produce that tensor. Thus you can access the mean/var attributes through that.

Pseudo code:
```python
>>> x_norm = batchnorm(x, gamma=g, beta=b, eps=1e-7)
>>> x_norm.creator
<mygrad.nnet.layers.batchnorm.BatchNorm at 0x22def8ab4e0>
>>> x_norm.creator.mean  # access the mean
np.array([...])
>>> x_norm.creator.var   # access the variance
np.array([...])

**Comment by IanCoolidge0** [comment]
The docstrings are updated and the commits should be squashed.

**Comment by davidmascharka** [comment]
Fantastic!
